src/ipc/StoreMap.cc

```diff
@@ -662,6 +671,100 @@ Ipc::StoreMap::validSlice(const int pos) const
     return 0 <= pos && pos < sliceLimit();
 }
 
+/// measures time and checks whether the elapsed time exceeds
+/// the given time interval
+class TimeMeter
+{
+    public:
+        explicit TimeMeter(const time_nsec_t max) : maxDuration(max) {
+            startTime = std::chrono::high_resolution_clock::now();
+        }
+
+        bool overflowed() {
+            if (!overflowed_) {
+                const auto endTime = std::chrono::high_resolution_clock::now();
+                const auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime);
+                overflowed_ = static_cast<time_nsec_t>(duration.count()) > maxDuration;
+            }
+            return overflowed_;
+        }
+
+    private:
+        std::chrono::time_point<std::chrono::high_resolution_clock> startTime;
+        const time_nsec_t maxDuration;
+        bool overflowed_ = false;
+};
+
+bool
+Ipc::StoreMap::validateHit(const sfileno fileno)
+{
+    const auto &anchor = anchorAt(fileno);
+
+    ++statCounter.hitValidation.attempts;
+
+    if (!anchor.lock.lockHeaders()) {
+        ++statCounter.hitValidation.refusalsDueToLocking;
+        return true; // presume valid; cannot validate changing entry
+    }
+
+    const uint64_t expectedByteCount = anchor.basics.swap_file_sz;
+
+    size_t actualSliceCount = 0;
+    uint64_t actualByteCount = 0;
+    SliceId lastSeenSlice = anchor.start;
+    TimeMeter timeMeter(Config.paranoid_hit_validation);
+    while (lastSeenSlice >= 0) {
+        ++actualSliceCount;
+        if (!validSlice(lastSeenSlice))
+            break;
+        const auto &slice = sliceAt(lastSeenSlice);
+        actualByteCount += slice.size;
+        if (actualByteCount > expectedByteCount)
+            break;
+        lastSeenSlice = slice.next;
+        if (timeMeter.overflowed())
+            break;
+    }
+
+    anchor.lock.unlockHeaders();
+
+    if (actualByteCount == expectedByteCount && lastSeenSlice < 0)
+        return true;
+
+    if (!anchor.basics.swap_file_sz) {
+        ++statCounter.hitValidation.refusalsDueToZeroSize;
+        return true; // presume valid; cannot validate w/o known swap_file_sz
+    }
+
+    if (timeMeter.overflowed()) {
```

Comment:
This is a bit wrong: The timer may have expired now, and not during the validation loop. The timeIsLimited changes sketched in another change request will fix this problem and eliminate overflowed_ data member as well.